src/Swarm/Game/Step.hs

```diff
@@ -1240,10 +1240,47 @@ execConst c vs s k = do
         inv <- use robotInventory
         return $ Out (VInt (fromIntegral $ countByName name inv)) s k
       _ -> badConst
+    Scout -> case vs of
+      [VDir d] -> do
+        rMap <- use robotMap
+        myLoc <- use robotLocation
+        heading <- deriveHeading d
+        botsByLocs <- use robotsByLocation
+        selfRid <- use robotID
+
+        -- Includes the base location, so we exclude the base robot later.
+        let locsInDirection :: [Location]
+            locsInDirection = take maxScoutRange $ iterate (.+^ heading) myLoc
+
+        let hasOpaqueEntity =
+              fmap (maybe False (`hasProperty` E.Opaque)) . entityAt
+
+        let hasVisibleBot :: Location -> Bool
+            hasVisibleBot = any botIsVisible . IS.toList . excludeSelf . botsHere
+             where
+              excludeSelf = (`IS.difference` IS.singleton selfRid)
+              botsHere loc = M.findWithDefault mempty loc botsByLocs
+              botIsVisible = maybe False canSee . (`IM.lookup` rMap)
+              canSee = not . (^. robotDisplay . invisible)
+
+        -- A robot on the same cell as an opaque entity is considered hidden.
+        let evalLocation :: Bool -> Location -> Maybe Bool
+            evalLocation isOpaque loc
+              | isOpaque = Just False
+              | hasVisibleBot loc = Just True
+              | otherwise = Nothing
+
+        let evalLocationM loc = do
+              opaque <- hasOpaqueEntity loc
+              return $ evalLocation opaque loc
+
+        result <- asum <$> mapM evalLocationM locsInDirection
```

Comment:
If I'm reading this right, I think this still has the property that it will always evaluate all 64 cells in the given direction no matter what?  It would be nice (though not critical) if it stops early as soon as it encounters either an opaque entity or a visible robot.

Reply:
Fixed by using `firstJustM`.